### DISCUSSION TITLE
What function or sequence of functions replace llama_sample_top_p_top_k?

### DISCUSSION THREAD

**Discussion by niansa** [ggml-org/llama.cpp#1351] [Q&A] [answered]
Hey!

If I just want to reproduce the behavior of the now-removed `llama_sample_top_p_top_k` function, what functions would I need and how would I call them?
I believe you'd have to call `llama_sample_top_k`, `llama_sample_top_p` and optionally `llama_sample_repetition_penalty` in sequence, but how'd I initialize `struct llama_token_data_array`?

Thanks
niansa

**Comment by olinorwell** [comment]
I'm also curious. I've come across that now removed function in code explaining something else in the discussions.

**Comment by Green-Sky** [comment, marked as answer]
check the changes to main.cpp https://github.com/ggerganov/llama.cpp/commit/dd7eff57d8491792010b1002b8de6a4b54912e5c
